core/src/lib/bsys.cc

```diff
@@ -417,92 +418,146 @@ int b_strerror(int errnum, char* buf, size_t bufsiz)
 }
 
 #if !defined(HAVE_WIN32)
-static bool del_pid_file_ok = false;
-#endif
-
-// Create a standard "Unix" pid file.
-void CreatePidFile(char* dir, const char* progname, int port)
+static void LockPidFile(const char* progname,
+                        int pidfile_fd,
+                        const char* pidfile_path)
 {
-#if !defined(HAVE_WIN32)
-  int pidfd = -1;
-  int len;
-  int oldpid;
-  char pidbuf[20];
-  POOLMEM* fname = GetPoolMemory(PM_FNAME);
-  struct stat statp;
+  struct flock fl;
+
+  fl.l_type = F_WRLCK;
+  fl.l_whence = SEEK_SET;
+  fl.l_start = 0;
+  fl.l_len = 0;
 
-  Mmsg(fname, "%s/%s.%d.pid", dir, progname, port);
-  if (stat(fname, &statp) == 0) {
-    /* File exists, see what we have */
-    *pidbuf = 0;
-    if ((pidfd = open(fname, O_RDONLY | O_BINARY, 0)) < 0
-        || read(pidfd, &pidbuf, sizeof(pidbuf)) < 0
-        || sscanf(pidbuf, "%d", &oldpid) != 1) {
+  if (fcntl(pidfile_fd, F_SETLK, &fl)) {
+    if (errno == EAGAIN || errno == EACCES) {
       BErrNo be;
-      Emsg2(M_ERROR_TERM, 0, _("Cannot open pid file. %s ERR=%s\n"), fname,
-            be.bstrerror());
+      Emsg2(M_ERROR_TERM, 0,
+            _("PID file '%s' is locked; probably '%s' is already running. "
+              "ERR=%s\n"),
+            pidfile_path, progname, be.bstrerror());
+
```

Comment:
I don't think the system error message for the locking failure adds any value here, maybe just don't mention it at all.
```suggestion
      Emsg2(M_ERROR_TERM, 0,
            _("PID file '%s' is locked; probably '%s' is already running.\n"),
            pidfile_path, progname);

```